### PR TITLE
Show failed files in web UI

### PR DESCRIPTION
**Pull request by calmh**
This introduces a new event `FolderErrors` that gets sent when we would log `puller not making progress`. That log entry has been demoted from warning to info. The web UI listens for this event and displays it in the folder status as `Errored Items`:

screen shot 2015-06-26 at 14 19 13

When clicked, it brings up a dialog with the item names (basename, with full in-folder path on hover) and errors (last error component, with full error text on hover):

screen shot 2015-06-26 at 14 19 25

(The "The/They" typo in the screenshot is fixed in the actual PR...)

Currently limited to 25 items. I considered pagination, but I'm not sure that's relevant - if there are 1000 failed files, probably they're all suffering from something similar and no one wants to investigate them individually...

Companion event docs: https://github.com/syncthing/docs/pull/22

### CODE REVIEW

**Comment by AudriusButkevicius** [comment]
You're a legend.

**Comment by uok** [comment]
Very helpful feature, thanks!
(maybe change "Errored Items" to "Sync errors" ?)

**Comment by calmh** [comment]
Yeah I think "Sync errors" and "Failed items" are better.

(@AudriusButkevicius :dancers:)

**Comment by plouj** [comment]
Would it be helpful to state that the list is limited to 25 files? Eg: "(Showing at most 25 items)"

**Comment by dcarosone** [comment]
Huzzah!

File: gui/index.html [hunk 1]
```diff
@@ -1027,6 +1035,21 @@ <h4 translate class="modal-title">Anonymous Usage Reporting</h4>
...
+  <modal id="failed" large="yes" status="warning" icon="exclamation-sign" close="yes" title="{{'Failed Items' | translate}}">
+    <p>
+      <span translate>The following items could not be synchronized.</span>
+      <span translate>They are retried automatically and will be synced when the error is resolved.</span>
+    </p>
+    <table class="table table-striped table-condensed">
+      <tr ng-repeat="e in currentFolderErrors">
+        <td><abbr title="{{e.path}}">{{e.path | basename}}</abbr></td>
+        <td><abbr title="{{e.error}}">{{e.error | lastErrorComponent}}</abbr></td>
+      </tr>
```
Comment (AudriusButkevicius) [inline gui/index.html]: So I still suggest pagination.

File: internal/model/rwfolder.go [hunk 2]
```diff
@@ -1402,3 +1432,57 @@ func moveForConflict(name string) error {
...
+	defer p.errorsMut.Unlock()
+
+	// We might get more than one error report for a file (i.e. error on
+	// Write() followed by Close()); we keep the first error as that is
+	// probably closer to the root cause.
+	for i := range p.errors {
+		if p.errors[i].Path == path {
+			return
+		}
+	}
```
Comment (AudriusButkevicius) [inline internal/model/rwfolder.go]: Why not just a map[string]error map?

Comment (AudriusButkevicius) [inline internal/model/rwfolder.go]: A `map[string]error` might fit here better?
You just check if it's present, and if it is, don't overwrite it.

**Comment by fatek** [comment]
I noticed the status of your 'default' folder still says 'Up to Date' even with the one errored item. It seems like users will only become aware of errored items after expanding each and every folder and checking for problems. Therefore, perhaps the folder status should change to 'Outdated' with the presence of at least one errored item.

**Comment by calmh** [comment]
Okay, feedback above adressed. Now shows state as "Out of Sync" when there are errors, at first with "..." for the errors (when the GUI has just loaded and we haven't received an event with the errors in it):

screen shot 2015-06-30 at 14 34 56

Then with the actual number of errors as a list with a link to the modal:

screen shot 2015-06-30 at 14 35 25

And the modal has pagination:

screen shot 2015-06-30 at 14 35 30

**Pull request merged**